pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -0,0 +1,183 @@
+	routingKey         = "mainfluxkey"
+	exchange           = "mainflux"
+	exchangeKind       = "fanout"
+	queueDurability    = true
+	queueDelete        = false
+	queueExclusivity   = false
+	queueWait          = false
+	consumerTag        = "mainflux-consumer"
+	mandatory          = false
+	immediate          = false
+)
+
+var (
+	errAlreadySubscribed = errors.New("already subscribed to topic")
+	errNotSubscribed     = errors.New("not subscribed")
+	errEmptyTopic        = errors.New("empty topic")
+)
+
+var _ messaging.PubSub = (*pubsub)(nil)
+
+// PubSub wraps messaging Publisher exposing
+// Close() method for RabbitMQ connection.
+type PubSub interface {
+	messaging.PubSub
+	Close()
+}
+
+type pubsub struct {
+	conn          *amqp.Connection
+	logger        log.Logger
+	queue         amqp.Queue
+	channel       *amqp.Channel
+	subscriptions map[string]bool
+	done          chan error
+	mutex         sync.Mutex
+}
+
+// NewPubSub returns RabbitMQ message publisher/subscriber.
+func NewPubSub(url, queueName string, logger log.Logger) (PubSub, error) {
+	endpoint := fmt.Sprintf("amqp://%s", url)
+	conn, err := amqp.Dial(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		return nil, err
+	}
+	queue, err := ch.QueueDeclare(queueName, queueDurability, queueDelete, queueExclusivity, queueWait, nil)
+	if err != nil {
+		return nil, err
+	}
+	ret := &pubsub{
+		conn:          conn,
+		queue:         queue,
+		channel:       ch,
+		logger:        logger,
+		subscriptions: make(map[string]bool),
+		done:          make(chan error),
+	}
+	return ret, nil
+}
+
+func (ps *pubsub) Publish(topic string, msg messaging.Message) error {
+	if topic == "" {
+		return errEmptyTopic
+	}
+	data, err := proto.Marshal(&msg)
+	if err != nil {
+		return err
+	}
+	subject := fmt.Sprintf("%s.%s.%s", exchange, chansPrefix, topic)
+	if err := ps.channel.ExchangeDeclare(subject, exchangeKind, true, false, false, false, nil); err != nil {
+		return err
+	}
+	err = ps.channel.Publish(
+		subject,
+		routingKey,
+		mandatory,
+		immediate,
+		amqp.Publishing{
+			Headers:     amqp.Table{},
+			ContentType: "text/plain",
+			Priority:    2,
+			AppId:       "mainflux",
+			Body:        []byte(data),
+		})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *pubsub) Subscribe(topic string, handler messaging.MessageHandler) error {
+	if topic == "" {
+		return errEmptyTopic
+	}
+	ps.mutex.Lock()
+	defer ps.mutex.Unlock()
+	if _, ok := ps.subscriptions[topic]; ok {
+		return errAlreadySubscribed
+	}
+
+	subject := fmt.Sprintf("%s.%s.%s", exchange, chansPrefix, topic)
+
+	if err := ps.channel.ExchangeDeclare(subject, exchangeKind, true, false, false, false, nil); err != nil {
+		return err
+	}
+
+	if err := ps.channel.QueueBind(ps.queue.Name, routingKey, subject, false, nil); err != nil {
+		return err
+	}
+	msgs, err := ps.channel.Consume(ps.queue.Name, "", true, false, false, false, nil)
```

Comment:
Is it OK to have an empty string for a consumer here?

Reply:
An empty string will cause the library to generate a unique identity for each consumer. Since we have no way to id the subscriber, I think that is the best fit for now.